core/src/gptTagStream.ts

```diff
@@ -0,0 +1,243 @@
+    }
+
+    private emitTagEvent(tag: Tag) {
+        this.emit("tag", tag);
+    }
+    private emitGeneratedEvent() {
+        this.emit("generated");
+    }
+
+    public stopGenerate(): void {
+        if (this.stream) {
+            try {
+                this.stream.destroy();
+            } catch (error) {
+                console.error('Failed to destroy the stream:', error);
+            } finally {
+                this.stream = null;
+            }
+        }
+    }
+
+    public isGenerating(): Boolean {
+        if (this.stream) {
+            return true
+        } else {
+            return false
+        }
+    }
+
+    private oldTags: Tag[] = []
+    public async generate(tags: Tag[], systemPrompt: string, remembrancePrompt: string) {
+
+        const apiKey = this.openaiConfig.apiKey;
+        const model = this.openaiConfig.model;
+
+        const configuration = new Configuration({ apiKey });
+        const openaiApi = new OpenAIApi(configuration);
+
+        const openaiStreamConfig = {
+            openai: openaiApi,
+            handler: {
+                onContent: (content: string, isFinal: boolean, stream: any) => {
+
+                    // console.log("🌵", content, "🌵");
+
+                    let newTags = this.extractTags(content);
+                    let diffTags = this.getUniqueTags(newTags, this.oldTags)
+                    this.oldTags = newTags;
+                    diffTags.forEach(diffTag => {
+                        this.emitTagEvent(diffTag)
+                    });
+                },
+                onDone: (stream: any) => {
+                    this.emitGeneratedEvent();
+                    this.stopGenerate();
+                },
+                onError: (error: Error, stream: any) => {
+                    // console.error("Openai Stream Error: ", error);
+                },
+            },
+        };
+
+        const messages = this.tagsToMessages(tags, systemPrompt, remembrancePrompt);
+        console.log("\n<🫥\n", messages, "\n🫥>\n")
+
+        const openaiStreamResponse = await OpenAIExt.streamServerChatCompletion(
+            {
+                model: model,
+                messages: messages,
+            },
+            openaiStreamConfig
+        );
+
+        this.stream = openaiStreamResponse.data;
+    }
+    
+    private getUniqueTags(newArray: Tag[], oldArray: Tag[]): Tag[] {
+        return newArray.filter((newTag) =>
+          !oldArray.some((oldTag) =>
+            newTag.role === oldTag.role && newTag.type === oldTag.type && newTag.text === oldTag.text
+          )
+        );
+      }
+
+
+    private tagsToMessages(tags: Tag[], systemPrompt: string, remembrancePrompt: string): OpenaiMessage[] {
+        // First, map each tag to an OpenaiMessage
+        const initialMessages = tags.map(tag => {
+            let content = tag.text;
+            if (tag.isRoleAssistant()) {
+                content = `<${tag.type}>${tag.text}</${tag.type}>`;
```

Comment:
```suggestion
                content = `<${tag.type}>
${tag.text}
</${tag.type}>`;
```
i think the newlines make it work better